### PR TITLE
Add optional slot swap

### PR DESCRIPTION
**Pull request by BurnedMarshal**
### List of changes
Add some improvements in deploy webapp pipeline:
- Add ping health check endpoint when a slot is present
- Add optional swap step configuration for selective deploy in production

### Motivation and context
The deploy pipeline has not a step that verify that the health check endpoint is alive after the deployment. A broken web app could be deployed causing production outages.
For some kind of deploy the team should have the possibility to deploy only on staging slots testing the application before the production switch. Moreover some deploy could be handled with progressive traffic routing between stage and productions slots.

### Type of changes

- [ ] Add new resources
- [X] Update configuration to existing resources
- [ ] Remove existing resources

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [X] No

### CODE REVIEW

File: .github/workflows/web_app_deploy.yaml [hunk 1]
```diff
@@ -22,6 +22,16 @@ on:
         type: boolean
         required: false
         default: true
+      health_check_path:
+        description: The health probe path exposed by the Web App.
+        type: string
+        default: ""
+        required: false
+      skip_swap_step:
```
Comment (gunzip) [inline .github/workflows/web_app_deploy.yaml:30]: what about an opt-intinstead of opt-out?

File: .github/workflows/web_app_deploy.yaml [hunk 2]
```diff
@@ -160,8 +170,17 @@ jobs:
            --type zip \
            --async false
 
+      - name: Ping Staging Health
+        if: ${{ inputs.use_staging_slot == 'true' && input.health_check_path != ''}}
```
Comment (arcogabbo) [inline .github/workflows/web_app_deploy.yaml:174]:
```suggestion
        if: ${{ inputs.use_staging_slot == 'true' && inputs.health_check_path != ''}}
```

File: .github/workflows/web_app_deploy.yaml [hunk 3]
```diff
@@ -160,8 +170,17 @@ jobs:
...
+        run: |
+          curl \
+            --retry 5 \
+            --retry-max-time 120 \
+            --retry-all-errors \
+            -f 'https://${{ inputs.web_app_name }}-staging.azurewebsites.net${{ inputs.health_check_path }}'
+
       - name: Swap Staging and Production Slots
-        if: ${{ inputs.use_staging_slot == true }}
+        if: ${{ inputs.use_staging_slot == true && input.skip_swap_step == false }}
```
Comment (arcogabbo) [inline .github/workflows/web_app_deploy.yaml:183]:
```suggestion
        if: ${{ inputs.use_staging_slot == true && inputs.skip_swap_step == false }}
```

**Review by arcogabbo — CHANGES_REQUESTED** [review]

**Comment by lucacavallaro** [comment]
We intentionally removed the ping from the health check because, with the right environment variables set (`WEBSITE_SWAP_WARMUP_PING_PATH` etc.), it becomes unnecessary.

See https://github.com/pagopa/dx/pull/47#discussion_r1645955762

**Comment by BurnedMarshal** [comment]
> We intentionally removed the ping from the health check because, with the right environment variables set (`WEBSITE_SWAP_WARMUP_PING_PATH` etc.), it becomes unnecessary.
> 
> See [#47 (comment)](https://github.com/pagopa/dx/pull/47#discussion_r1645955762)

Thanks for the information, I miss it. Unfortunately this pipeline has an hidden requirement to work properly and safely. If I'm not wrong only 2 function has the correct settings in their configuration `fims` and `cgn-onboarding-portal-infra`. We should think a way to inform clearly the final users about all the requirements.
For now I'm removing the ping step and the related variable, adding a small comment with the Azure documentation reference.

File: .github/workflows/web_app_deploy.yaml (hunk 1 above)
Comment (BurnedMarshal) [reply .github/workflows/web_app_deploy.yaml]: Fixed in [65c6818](https://github.com/pagopa/dx/pull/77/commits/65c6818bd914ae9bb60cc32ceada88b62996a63b)

**Comment by lucacavallaro** [comment]
> > We intentionally removed the ping from the health check because, with the right environment variables set (`WEBSITE_SWAP_WARMUP_PING_PATH` etc.), it becomes unnecessary.
> > See [#47 (comment)](https://github.com/pagopa/dx/pull/47#discussion_r1645955762)
> 
> Thanks for the information, I miss it. Unfortunately this pipeline has an hidden requirement to work properly and safely. If I'm not wrong only 2 function has the correct settings in their configuration `fims` and `cgn-onboarding-portal-infra`. We should think a way to inform clearly the final users about all the requirements. For now I'm removing the ping step and the related variable, adding a small comment with the Azure documentation reference.

maybe we can update the function/app services modules to include these variables by default

File: .github/workflows/web_app_deploy.yaml [hunk 4]
```diff
@@ -22,6 +22,11 @@ on:
         type: boolean
         required: false
         default: true
+      swap_slot_step:
+        description: Skip slot swap for selective deployment and testing.
```
Comment (gunzip) [inline .github/workflows/web_app_deploy.yaml:26]:
```suggestion
        description: Swap production and staging slot.
```

**Comment by gunzip** [comment]
During our latest meeting, we discussed optimizing our code testing process using slots. We decided to experiment with preview slots—assigning one to each pull request—before finalizing our approach. This should help us determine the most effective strategy moving forward. What do you think about this @BurnedMarshal ?

**Pull request closed without merging**